cedar-drt/fuzz/src/schemas.rs

```diff
@@ -331,53 +463,58 @@ fn action_type_equivalence<N: PartialEq + Debug + Display + Clone + TypeName + O
     }
 }
 
-/// ApplySpecs are equivalent iff
-/// A) the principal and resource type lists are equal
-/// B) the context shapes are equivalent
-fn apply_spec_equiv<N: TypeName + Clone + PartialEq + Ord>(
-    lhs: &ApplySpec<N>,
-    rhs: &ApplySpec<N>,
-) -> bool {
-    shape_equiv(&lhs.context.0, &rhs.context.0)
-        && vector_equiv(&lhs.principal_types, &rhs.principal_types)
-        && vector_equiv(&lhs.resource_types, &rhs.resource_types)
+impl<N: TypeName + Clone + PartialEq + Ord + Debug + Display> Equiv for ApplySpec<N> {
+    fn equiv(lhs: &Self, rhs: &Self) -> Result<(), String> {
+        // ApplySpecs are equivalent iff
+        // A) the principal and resource type lists are equal
+        // B) the context shapes are equivalent
+        if Equiv::equiv(&lhs.context.0, &rhs.context.0).is_ok()
+            && vector_equiv(&lhs.principal_types, &rhs.principal_types)
+            && vector_equiv(&lhs.resource_types, &rhs.resource_types)
+        {
+            Ok(())
+        } else {
+            Err("ApplySpecs are not equivalent".into())
+        }
+    }
 }
 
 fn either_empty<N>(spec: &json_schema::ApplySpec<N>) -> bool {
     spec.principal_types.is_empty() || spec.resource_types.is_empty()
 }
 
-/// Just compare entity attribute types and context types are equivalent
-pub fn validator_schema_attr_types_equivalent(
-    schema1: &cedar_policy_validator::ValidatorSchema,
-    schema2: &cedar_policy_validator::ValidatorSchema,
-) -> bool {
-    let entity_attr_tys1: HashMap<
-        &cedar_drt::ast::EntityType,
-        HashMap<&smol_str::SmolStr, &cedar_policy_validator::types::AttributeType>,
-    > = HashMap::from_iter(
-        schema1
-            .entity_types()
-            .map(|(name, ty)| (name, HashMap::from_iter(ty.attributes()))),
-    );
-    let entity_attr_tys2 = HashMap::from_iter(
-        schema2
-            .entity_types()
-            .map(|(name, ty)| (name, HashMap::from_iter(ty.attributes()))),
-    );
-    let context_ty1: HashSet<&cedar_policy_validator::types::Type> = HashSet::from_iter(
-        schema1
-            .action_entities()
-            .unwrap()
-            .iter()
-            .map(|e| schema1.get_action_id(e.uid()).unwrap().context_type()),
-    );
-    let context_ty2: HashSet<&cedar_policy_validator::types::Type> = HashSet::from_iter(
-        schema2
-            .action_entities()
-            .unwrap()
-            .iter()
-            .map(|e| schema1.get_action_id(e.uid()).unwrap().context_type()),
-    );
-    entity_attr_tys1 == entity_attr_tys2 && context_ty1 == context_ty2
+impl Equiv for cedar_policy_validator::ValidatorSchema {
+    fn equiv(lhs: &Self, rhs: &Self) -> Result<(), String> {
```

Comment:
Seems a little weird that this function isn't using any other `equiv` functions (e.g., for `HashMap`s)